golos.publication/golos.publication.cpp

```diff
@@ -143,69 +142,79 @@ void publication::delete_post(account_name account, std::string permlink) {
     }
 }
 
-void publication::upvote(account_name voter, account_name author, std::string permlink, asset weight) {
+void publication::upvote(account_name voter, account_name author, std::string permlink, int64_t weight) {
     require_auth(voter);
     recovery_battery(voter, &structures::account_battery::limit_battery_of_votes, {UPPER_BOUND, VOTE_OPERATION_INTERVAL, UPPER_BOUND, type_recovery::linear});
 
-//    eosio_assert(weight > 0, "The weight sign can't be negative.");
+    eosio_assert(weight > 0, "The weight sign can't be negative.");
 
     set_vote(voter, author, permlink, weight);
 }
 
-void publication::downvote(account_name voter, account_name author, std::string permlink, asset weight) {
+void publication::downvote(account_name voter, account_name author, std::string permlink, int64_t weight) {
     require_auth(voter);
 
-//    eosio_assert(weight < 0, "The weight sign can't be positive.");
+    eosio_assert(weight > 0, "The weight sign can't be negative.");
 
-    set_vote(voter, author, permlink, weight);
+    set_vote(voter, author, permlink, -weight);
 }
 
 void publication::unvote(account_name voter, account_name author, std::string permlink) {
     require_auth(voter);
-
-//    eosio_assert(weight == 0, "The weight can be only zero.");
-
-    set_vote(voter, author, permlink, 0);
+  
+    set_vote(voter, author, permlink, UNVOTE_WEIGHT);
 }
 
-void publication::close_post() {
-    close_post_timer();
+void publication::close_post(account_name account, std::string permlink) {
+    tables::post_table post_table(_self, account);
+    checksum256 permlink_hash;
+    sha256(permlink.c_str(), sizeof(permlink), &permlink_hash);
+    auto posttable_index = post_table.get_index<N(permlink)>();
+    auto posttable_obj = posttable_index.find(structures::post::get_hash_key(permlink_hash));
+
+    posttable_index.modify(posttable_obj, account, [&]( auto &item) {
+        item.isclosed = POST_CLOSED;
+    });
 }
 
-void publication::close_post_timer() {
+void publication::close_post_timer(account_name account, std::string permlink) {
     require_auth(_self);
 
     transaction trx;
-    trx.actions.emplace_back(action{permission_level(_self, N(active)), _self, N(closepost), structures::st_hash{now()}});
-    trx.delay_sec = CLOSE_POST_TIMER;
+    trx.actions.emplace_back(action{permission_level(_self, N(active)), _self, N(closepost), structures::closepost{account, permlink}});
+    trx.delay_sec = CLOSE_POST_PERIOD;
     trx.send(_self, _self);
 }
 
+// weight is fixed point with two decimal digits,
+// if you need to operate it,
+// please divide by TWO_DECIMAL_DIGITS
 void publication::set_vote(account_name voter, account_name author,
-                           std::string permlink, asset weight) {
-    tables::vote_table vote_table(_self, author);
+                           std::string permlink, int64_t weight) {
```

Comment:
weight here is percent. `int16_t` is enough to hold it, it can be in range from –10000 to 10000 (equal to –100,00% to 100,00%)